### DISCUSSION TITLE
Unclear how to deserialize a form within a request body

### DISCUSSION THREAD

**Discussion by brassy-endomorph** [actix/actix-web#3113] [Q&A]
The current patterns is to use a `FormConfig` and then the `Form<T>` extractor. The config has an err handler function, but that's not remotely close to what I want. If a user POSTs a form, I want to be able to read the form and render the page with errors under each form field. I can get access to these errors by writing a custom deserializer for my struct (fine, but annoying), but there's no apparent way to get this behavior one has with Python + Flask + WTForms.

```python
@app.route("/", methods=["GET", "POST"])
def my_handler(request):
    form = MyForm()
    if form.validate_on_submit():
        # do stuff with form data, save to DB, etc.,
        return redirect(url_for('another_endpoint'))
    return render_template('index.html', form=form)
```

This means I have one place I register the page on the original GET and the POST if the data is invalid. Actix-web moves my handler somewhere extremely far away if I follow the expected pattern like:

```rust
#[routes]
#[get("/")]
#[post("/")]
async fn my_handler(form: web::Form<MyForm>) -> Result<String> {
   todo!("Can't access form errors from here")
}
```

So you'd think I can do something like this:

```rust
async fn my_handler(req: HttpRequest) -> Result<String> {
    let form = Form::<MyForm>::from_req(req, req.take_payload());
    // do something with the form
}
```

```
error[E0308]: mismatched types
   |
   |    let form = Form::<MyForm>::from_req(req, &mut req.take_payload());
   |               ------------------------      ^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut Payload<Pin<Box<dyn Stream<Item = Result<Bytes, ...>>>>>`, found `Payload<()>`
   |                |
   |                arguments to this function are incorrect
   |
   = note: expected mutable reference `&mut actix_web::dev::Payload`
              found mutable reference `&mut actix_web::dev::Payload<()>`
```

Except I get type errors from `take_payload` and there's not a clear way to coerce the types around. What is the incantation I need to do so that I can deserialize a form within a handler?

**Comment by robjtede** [comment]
If the request is well-formed, even if invalid from a business logic perspective then, `form: web::Form<MyForm>` will allow you to do more validation on it within the handler. This is how I'd recommend you proceed acquiring the form data.

After that, the [`validator`](https://docs.rs/validator) crate is a good crate for adding more requirements. A pattern to use for required fields is to set all fields as `Option<T>` and adding a `#[validate(required)]` annotation. Using this technique will collect _all_ the form errors and provide more flexibility with custom validation rules.

(For completeness, I want to point out that `actix_web::Result<Form<T>>` is a possible extractor, and would provide an error to your handler if one was raised during deserialization of the request payload. However, I would not include any field-level error information that you could extract, so it's not suitable for your problem as I understand it.)
